src/java.base/share/classes/java/net/JarURLConnection.java

```diff
@@ -152,44 +155,63 @@ public abstract class JarURLConnection extends URLConnection {
      * could be found in a specification string or the
      * string could not be parsed.
      */
-
     protected JarURLConnection(URL url) throws MalformedURLException {
         super(url);
-        parseSpecs(url);
-    }
 
-    /* get the specs for a given url out of the cache, and compute and
-     * cache them if they're not there.
-     */
-    private void parseSpecs(URL url) throws MalformedURLException {
+        // Extract JAR file URL and entry name components from the URL
         String spec = url.getFile();
-
         int separator = spec.indexOf("!/");
-        /*
-         * REMIND: we don't handle nested JAR URLs
-         */
+
+        // REMIND: we don't handle nested JAR URLs
         if (separator == -1) {
             throw new MalformedURLException("no !/ found in url spec:" + spec);
         }
 
-        @SuppressWarnings("deprecation")
-        var _unused = jarFileURL = new URL(spec.substring(0, separator++));
+        jarFileURL = parseJarFileURL(spec, separator, url);
+        entryName = parseEntryName(spec, separator);
+    }
 
+    /**
+     * Parse the URL of the JAR file backing this JarURLConnection,
+     * appending any #runtime fragment as neccessary
+     *
+     * @param spec the URL spec of this connection
+     * @param separator the index of the '!/' separator
+     * @param connectionURL the URL passed to the constructor
+     * @return a URL to the JAR file this connection reads from
+     *
+     * @throws MalformedURLException if a malformed URL is found
+     */
+    @SuppressWarnings("deprecation")
+    private static URL parseJarFileURL(String spec, int separator, URL connectionURL) throws MalformedURLException {
```

Comment:
Nit - maybe rename `separator` to `separatorIndex`? Here and in the other private method.

Reply:
Thanks Jai, I have renamed 'separator' as suggested, and also 'nameIdx' for consistency.

tier1 runs on GHA, will run tier2 locally.